dyno: Always use the dyno parser

dlongnecke-cray opened the pull request:
dyno: Always use the dyno parser (#19816)

Always use the dyno parser to parse Chapel source files. Do this
by setting `fDynoCompilerLibrary` to `true`.

Adjust ~20 `.good` files to accept the output produced by the dyno
parser. These differences are small, mostly off-by-one on line
numbers (e.g., the old parser would pin a function on the location
of its first formal, but we use the location of the function's name).

Adjust ~16 `.bad` files to accept new outputs.

Adjust the converter to handle attaching chpldoc comments.

Add a function to the converter 'trimBlockCommentForDocs' which
duplicates the work in the old scanner's 'processBlockComment'
for doc comments.

Adjust the converter to take in the builder result so that it can
fetch the location of comments.

Remove the `test/dyno` directory.

TESTING

- [x] `ALL` on `linux64` when `COMM=none`, `LLVM=system`
- [x] `ALL` on `linux64` when `COMM=gasnet`, `LLVM=system`
- [x] `ALL` on `linux64` when `COMM=none`, `LLVM=none`
- [x] `linux64` when `COMM=none`, `LLVM=system`, `-futures-only`

Reviewed by @mppf. Thanks!

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

bradcray commented:
High-level approach question:  Does it make sense to remove the old parser immediately rather than waiting until we have our feet under us with the new one?  (I'd been imagining that in the same way one could throw `--dyno` to get the new parser today, there might be a way to throw `--old-parser` to keep using the old one).  Or does that carry some burden with it that I'm not anticipating?

dlongnecke-cray commented:
If anything goes catastrophically wrong and we can't fix it within a day, we plan to back out this PR.

It is true that devs will need to get used to the new parser, but we'll have to do that at some point anyway and we're ready to help.

I think there are a few burdens involved in carrying the old parser around. 

- Confidence - I think we (myself at least) were never quite confident that dyno's parser was ready until this PR. For example, details of chpldoc are deeply embedded in the old parser, and we did not iron out those bugs for dyno until we removed it.
- Maintenance - we don't want to have to maintain two parsers (already in the past month we've had two parser changes that we've had to propagate to dyno)
- Integration - there are a lot of diverging paths in the codebase today, and we've added more and more related to dyno functionality, and it would be nice to collapse a bunch of those as we head into "dyno-as-resolver" land

What benefit do you see from keeping the old parser around?

bradcray commented:
> What benefit do you see from keeping the old parser around?

* Conservatism in the rock-climbing sense (don't take your weight off something until you're sure your new grip is firm)
* The ability, should anything go wrong, to check whether it's due to the old vs. new parser from a single source base
* The potential ability to migrate branches that I have that modify parse-time code incrementally rather than throwing them out and starting from scratch

With respect to that last question:  what is the role of build.cpp with the new parser?  I was imagining more of it going away if the old parser did, but based on the diff, it seems that's not the case.  Does the lowering from uAST to old IR use build.cpp to do so?

mppf commented on Makefile:
```diff
@@ -79,10 +79,6 @@ compiler: FORCE
 	@cd third-party && $(MAKE) CHPL_MAKE_HOST_TARGET=--host jemalloc
 	@cd compiler && $(MAKE)
 
-parser: FORCE
```
I think this should just do `make dyno-parser`

mppf commented on compiler/passes/convert-uast.cpp:
```diff
@@ -1129,7 +1267,7 @@ struct Converter {
     if (node->iterand()->isZip()) return false;
     if (node->numStmts() != 1) return false;
-    if (node->index()) return false;
+    if (node->index() && node->stmt(0)->isConditional()) return false;
```
Is this something like `[i in 1..100]`? How could that be an array type?

dlongnecke-cray replied:
It may not be an array type, but the old parser considers it one today in some cases. The problem is the AST printer uses `maybeArrayType` to print out square braces. So if we change that we can change the logic here.

mppf replied:
OK, in that case please just add a comment saying that same thing

mppf commented on compiler/passes/convert-uast.cpp:
```diff
@@ -1844,6 +1983,13 @@ struct Converter {
+        INT_ASSERT(defExpr);
+        auto varSym = toVarSymbol(defExpr->sym);
+        INT_ASSERT(varSym);
+        varSym->doc = comment;
```
Won't this attach the comment to all of the MultiDecl variables? Is that what is intended? I would think it would only go on the 1st of them.  You could do that by setting `comment = nullptr;` after using the comment here.

dlongnecke-cray replied:
I think in `buildVarDecls` we loop through and attach the doc comment to each `VarSymbol`, it would be nice to get a double check on that though. @lydia-duncan ?

lydia-duncan replied:
In practice, I suspect multi-variable declarations do not turn up much in documentation in practice - if the user wants them to have different documentation, they'll separate into single declarations; if the user wants the same documentation, they wouldn't have any recourse except copying unless we do as the code is currently written

mppf commented on compiler/passes/convert-uast.cpp:
```diff
@@ -1881,11 +2029,22 @@ struct Converter {
+          auto varSym = toVarSymbol(defExpr->sym);
+          INT_ASSERT(varSym);
+          varSym->doc = comment;
+        }
```
Same here

dlongnecke-cray commented:
> With respect to that last question: what is the role of build.cpp with the new parser?

The builders do an often substantial amount of lowering (some things are lowered from the getgo as they are parsed), and the lowered form is sometimes critical to further passes functioning correctly, so we cannot remove them yet. We would like to gradually remove all semantic checks from the builders until all they are doing is lowering, though.

As we proceed with resolution from dyno I imagine we will begin to a take a more hands on approach and modify/refactor the builders more extensively.

mppf commented:
@bradcray 

> > What benefit do you see from keeping the old parser around?
> 
>     * Conservatism in the rock-climbing sense (don't take your weight off something until you're sure your new grip is firm)

Either things go OK in testing or they don't and we have to revert it. I don't see why removing or not removing the old parser changes that.

>     * The ability, should anything go wrong, to check whether it's due to the old vs. new parser from a single source base

That is an interesting capability. Would you imagine we would want to release that way? I'm hoping we can get out of maintaining 2 parsers as soon as possible. In terms of nightly testing etc. I am expecting it will be pretty clear from error messages if it was a parsing issue.

>     * The potential ability to migrate branches that I have that modify parse-time code incrementally rather than throwing them out and starting from scratch

I think we're going to want to remove the old parser at some point and I don't think it's a given that all not-yet-merged changes to the old parser will have been merged by that time, whenever that is. So while we could put it off a few weeks, if these are branches that aren't under active development, I don't see how that helps much. So I suppose it depends on when you plan to get back to those branches. But anyway, if you needed to port parser changes over to dyno, you could do that with a version of `main` just before the old parser is removed, in order to be able to do it incrementally.

> With respect to that last question: what is the role of build.cpp with the new parser? I was imagining more of it going away if the old parser did, but based on the diff, it seems that's not the case. Does the lowering from uAST to old IR use build.cpp to do so?

Yes. We are using build.cpp helpers to construct the old AST from uAST in the converter.

@bradcray - supposing that we did not remove the old parser in this PR, when do you think we would do that, given that it costs something to maintain two parsers and code paths to call them?

bradcray commented:
> supposing that we did not remove the old parser in this PR, when do you think we would do that, given that it costs something to maintain two parsers and code paths to call them?

To clarify, I'm not imagining that we would continue to develop the old parser, simply to let it be used as-is of the day of the switch if needed / desired.  I'm not sure when we'd remove the old parser but expect it would depend a lot on the experience we have with the new one (e.g., whether in a week or a month people were like "I've never thought about or used the old parser since the switch" vs. "I've continued to find it useful from time to time for the following reasons."

dlongnecke-cray commented:
Ok, we can look at adding an `--old-parser` flag and delaying removing the old parser for another PR. It might take a little bit of time, though.

dlongnecke-cray commented:
One thing I am worried about is hardening all the places where we branch around dyno or not to make sure we are using the old/new parser appropriately, for instance we didn't actually discover all the `chpldoc` failures until after we removed the old parser completely, for some reason I'm not sure of yet. Might be that we invoke the parser in a different way.

There are also a few more branches I know I'll have to introduce to make sure everything is toggled on.

lydia-duncan commented:
Personally, I think we've been living in the two parser world long enough that I'd be comfortable shifting over to just the new one.  It's been easy enough to add similar changes to the new one when I've added something that I don't feel it will be a steep learning curve and I'd like to not have to think about both any more and clear the cache of the old one from my brain

bradcray commented:
> It's been easy enough to add similar changes to the new one when I've added something

To be clear, I'm not proposing we maintain the old one so that people can develop on it and avoid the new one (which wouldn't actually work since testing would be running the new one), but so that it can still be run on demand or if/when desired.  Again, my thinking was "Both are supported today and we can choose between them" so it seemed like the most conservative step would be to continue to support both but to flip the default.

bradcray commented:
Different comment (just because I just ran into this test in my own paratest):  Can `test/dyno/primers` be removed in this PR?  (I don't think it is anyway...)

bradcray commented:
> we can look at adding an --old-parser flag

It looks like we already have a `--no-dyno` flag.  Rather than adding a new flag, could the approach be as simple as swapping the default for the existing flag?  Or does `--dyno` imply additional things other than "use the new parser" which would make  it inappropriate?

mppf commented:
> > we can look at adding an --old-parser flag
> 
> It looks like we already have a `--no-dyno` flag. Rather than adding a new flag, could the approach be as simple as swapping the default for the existing flag? Or does `--dyno` imply additional things other than "use the new parser" which would make it inappropriate?

It would be OK right now but after we get the parser swapped in we are expecting `--dyno` to activate the new scope resolver / function resolver.

dlongnecke-cray commented:
I would like to advocate a bit more strongly for just doing the swap in one go. I was intending to migrate semantic checks out of the old parser builders and potentially remove/noop the `checkParsed` pass. As long as the old parser remains in place I don't think I can remove anything.

I think we would want to reclaim `--dyno` for use with the new resolver, so we'd need to add a different `--old-parser` flag.

Additionally, as soon as someone needs to touch the parser I think we lose most of the benefit of carrying it around, as we either have to update both of them or risk one being out of sync with the other.

dlongnecke-cray commented:
There were changes to 16 `.bad` files for futures. Taking a look at them now.

dlongnecke-cray commented:
This is almost ready to merge. Just to reaffirm, the new direction is to keep the old parser around (for now). It can be used by throwing `--no-dyno`.
